pkg/controller/clone-controller.go

```diff
@@ -0,0 +1,264 @@
+	}
+	// see if pvc's pod phase anno needs to be added/updated
+	phase := string(pod.Status.Phase)
+	if !checkIfAnnoExists(pvc, AnnPodPhase, phase) {
+		pvc, err = setPVCAnnotation(c.clientset, pvc, AnnPodPhase, phase)
+		if err != nil {
+			return errors.WithMessage(err, fmt.Sprintf("could not set annotation \"%s: %s\" on pvc %q", AnnPodPhase, phase, pvc.Name))
+			glog.V(Vdebug).Infof("processPodItem: pod phase %q annotated in pvc %q", pod.Status.Phase, pvcKey)
+		}
+	}
+	return nil
+}
+
+// Select only pvcs with the 'CloneRequest' annotation and that are not being processed.
+// We forget the key unless `processPvcItem` returns an error in which case the key can be
+// retried.
+func (c *CloneController) ProcessNextPvcItem() bool {
+	key, shutdown := c.pvcQueue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.pvcQueue.Done(key)
+
+	pvc, err := c.pvcFromKey(key)
+	if err != nil || pvc == nil {
+		return c.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: error converting key %v to pvc: %v", key, err))
+	}
+	// filter pvc and decide if the source and target pods should be created
+	if continue_processing, _, _ := checkClonePVC(c.clientset, pvc, false); !continue_processing {
+		return c.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: skipping pvc %q\n", key))
+	}
+	glog.V(Vdebug).Infof("ProcessNextPvcItem: next pvc to process: %s\n", key)
+	err = c.processPvcItem(pvc)
+	if err != nil { // processPvcItem errors may not have been logged so log here
+		glog.Errorf("error processing pvc %q: %v", key, err)
+		return true
+	}
+	return c.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: processing pvc %q completed", key))
+}
+
+// Create the cloning source and target pods based the pvc. The pvc is checked (again) to ensure that we are not already
+// processing this pvc, which would result in multiple pods for the same pvc.
+func (c *CloneController) processPvcItem(pvc *v1.PersistentVolumeClaim) error {
+	cr, err := getCloneRequestPVC(pvc)
+	if err != nil {
+		return err
+	}
+	
+	// check our existing pvc one more time to ensure we should be working on it
+	// and to help mitigate any race conditions. This time we get the latest pvc.
+	doCreate, pvc, err := checkClonePVC(c.clientset, pvc, true)
+	if err != nil { // maybe an intermittent api error
+		return err
+	}
+	if !doCreate { // don't create pods but not an error
+		return nil // forget key; logging already done
+	}
+
+secret := ""
+	// all checks passed, let's create the source pod!
+	_, err = CreateCloneSourcePod(c.clientset, c.cloneImage, c.verbose, c.pullPolicy, cr, secret, pvc)
+	if err != nil {
+		return err
+	}
+	
+	//create the target pod
+	_ , err = CreateCloneTargetPod(c.clientset, c.cloneImage, c.verbose, c.pullPolicy, cr, secret, pvc)
+	if err != nil {
+		return err
+	}
+	
+	// update pvc with 'CloneOf' annotation optional cdi label
+	anno := map[string]string{AnnCloneInProgress: "true"}
+	var lab map[string]string
+	pvc, err = updatePVC(c.clientset, pvc, anno, lab)
```

Comment:
Do we remove/update this annotation when the cloning finishes (either with success or failure?

Reply:
I changed the name from In-progress to clonePods (in importer pod its import-pod), so it will not confuse us when pods complete. We can talk about it of course...